### ISSUE TITLE
validate()

### ISSUE THREAD

**Issue by aliparsai** [AlexandreDecan/sismic#30]
validate() method was really handy for mutation, since I'm changing the internals of the mutants, and it was nice to have  a method that detects inconsistencies. Is there anything in place of it that I can use instead?

**Comment by AlexandreDecan** [comment]
The new ``model.Statechart``api makes all the validation during construction. 
You can:
 - ``add_state``
 - ``remove_state``
 - ``rename_state``
 - ``add_transition``
 - ``remove_transition``
 - ``rotate_transition``

The list of states is available with ``states`` (returns names, use ``state_for(name)`` after) while the list of transitions is available with ``transitions``.

These methods raise ``exceptions.StatechartError`` if something goes wrong.

**Comment by aliparsai** [comment]
I think having a separate (possibly redundant) validation method is very useful. For example, when the statechart instances are produced by something other than the importers included. This way, one can check the consistency of the model instance without assuming correct behaviour by the creator of the model instance.

**Comment by AlexandreDecan** [comment]
The validation already occurs when you manually construct a ``Statechart``, and does not depend on any importer. The only "old check" that is not done during the construction of a statechart is to check that initial states and memories refer to an existing parent's substate. I do not see how to handle this as a statechart is created using a top-down approach. 

I'm looking toward a solution.

**Comment by AlexandreDecan** [comment]
Consider looking at 7b38ad0. I implemented the two checks (that can also be called independently).
Notice that ``remove_state`` now allows a state to be removed even if it has nested states, if it is the target or the source of a transition, or if it is the target of a composite state's initial value or history state's memory value.

**Comment by aliparsai** [comment]
Good, that makes things much easier.

**Comment by AlexandreDecan** [comment]
Take care, the current implementation of ``rename_state`` is buggy and does not ``validate()``. 
I'm working on it (next push will contain the patch)

**Comment by AlexandreDecan** [comment]
The same bug occurred when a state was removed. FYI, it is now fixed. 

The internal stuff is more consistent: ``Statechart._children`` and ``Statechart._parent`` have an entry for every state, even if this state has no child (for ``_children``) or no parent (for ``_parent``). 

I added several tests (including ``self.validate()`` after each structural manipulation). I hope this is sufficient, but I'm confident.
